uamqp/error.py

```diff
@@ -7,10 +7,133 @@
 from enum import Enum
 from collections import namedtuple
 
-from .constants import PORT, FIELD
+from .constants import SECURE_PORT, FIELD
 from .types import AMQPTypes, FieldDefinition
 
 
+class ErrorCodes(bytes, Enum):
+    InternalError = b"amqp:internal-error"
+    IllegalState = b"amqp:illegal-state"
+    DecodeError = b"amqp:decode-error"
+    NotFound = b"amqp:not-found"
+    NotImplemented = b"amqp:not-implemented"
+    NotAllowed = b"amqp:not-allowed"
+    InvalidField = b"amqp:invalid-field"
+    ResourceLocked = b"amqp:resource-locked"
+    ResourceDeleted = b"amqp:resource-deleted"
+    UnauthorizedAccess = b"amqp:unauthorized-access"
+    FrameSizeTooSmall = b"amqp:frame-size-too-small"
+    ResourceLimitExceeded = b"amqp:resource-limit-exceeded"
+    PreconditionFailed = b"amqp:precondition-failed"
+    ConnectionRedirect = b"amqp:connection:redirect"
+    ConnectionCloseForced = b"amqp:connection:forced"
+    ConnectionFramingError = b"amqp:connection:framing-error"
+    SessionWindowViolation = b"amqp:session:window-violation"
+    SessionErrantLink = b"amqp:session:errant-link"
+    SessionHandleInUse = b"amqp:session:handle-in-use"
+    SessionUnattachedHandle = b"amqp:session:unattached-handle"
+    LinkRedirect = b"amqp:link:redirect"
+    LinkStolen = b"amqp:link:stolen"
+    LinkDetachForced = b"amqp:link:detach-forced"
+    LinkTransferLimitExceeded = b"amqp:link:transfer-limit-exceeded"
+    LinkMessageSizeExceeded = b"amqp:link:message-size-exceeded"
+    ClientError = b"amqp:client-error"
+    UnknownError = b"amqp:unknown-error"
+    VendorError = b"amqp:vendor-error"
+    SocketError = b"amqp:socket-error"
+
+
+class RetryMode(str, Enum):
+    Exponential = 'exponential'
+    Fixed = 'fixed'
+
+
+class ErrorPolicy:
+
+    no_retry = [
+        ErrorCodes.DecodeError,
+        ErrorCodes.LinkMessageSizeExceeded,
+        ErrorCodes.NotFound,
+        ErrorCodes.NotImplemented,
+        ErrorCodes.LinkRedirect,
+        ErrorCodes.NotAllowed,
+        ErrorCodes.UnauthorizedAccess,
+        ErrorCodes.LinkStolen,
+        ErrorCodes.ResourceLimitExceeded,
+        ErrorCodes.ConnectionRedirect,
+        ErrorCodes.PreconditionFailed,
+        ErrorCodes.InvalidField,
+        ErrorCodes.ResourceDeleted,
+        ErrorCodes.IllegalState,
+        ErrorCodes.FrameSizeTooSmall,
+        ErrorCodes.ConnectionFramingError,
+        ErrorCodes.SessionUnattachedHandle,
+        ErrorCodes.SessionHandleInUse,
+        ErrorCodes.SessionErrantLink,
+        ErrorCodes.SessionWindowViolation
+    ]
+
+    def __init__(
+        self,
+        **kwargs
+    ):
+        """
+
+        keyword int retry_total:
+        keyword float retry_backoff_factor:
+        keyword float retry_backoff_max:
+        keyword RetryMode retry_mode:
+        keyword list no_retry_condition:
+        keyword dict custom_retry_policy:
```

Comment:
draft: custom_retry_policy should be function,

overload retry_backoff_factor it could either be a float/callable returns a float

condition a,b,c,
condition d,e,f microsoft codes, 4 seconds

input: it's taking the AMQPException
output: float